### PR TITLE
Make "mergeIntoParent" information available during compile of semantics tree

### PR DESCRIPTION
**Pull request by goderbauer**
Previously, the information if a node is merged into its parent was only calculated after the shape of the semantics tree had been updated as the very last step before sending the updated tree to the engine. With this change, the information is calculated on-the-fly while the tree is compiled.

We need the information during the tree compile because in combination with the clip rect we can then decide on-the-fly to skip compiling those parts of the tree that will not be visible on screen. This will be implemented in a separate PR.

### CODE REVIEW

File: packages/flutter/lib/src/rendering/object.dart [hunk 1]
```diff
@@ -787,7 +787,12 @@ class _ConcreteSemanticsFragment extends _InterestingSemanticsFragment {
     SemanticsAnnotator annotator,
     Iterable<_SemanticsFragment> children,
     bool dropSemanticsOfPreviousSiblings,
-  }) : super(renderObjectOwner: renderObjectOwner, annotator: annotator, children: children, dropSemanticsOfPreviousSiblings: dropSemanticsOfPreviousSiblings);
+    bool mergeIntoParent,
+    bool mergesAllDescendants,
+  }) : _mergeIntoParent = mergeIntoParent, _mergesAllDescendants = mergesAllDescendants, super(renderObjectOwner: renderObjectOwner, annotator: annotator, children: children, dropSemanticsOfPreviousSiblings: dropSemanticsOfPreviousSiblings);
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/object.dart]: probably time to wrap this line :-)

File: packages/flutter/lib/src/rendering/object.dart [hunk 2]
```diff
@@ -2678,14 +2696,17 @@ abstract class RenderObject extends AbstractNode with DiagnosticableTreeMixin im
...
   /// appropriate [_SemanticsFragment] object that describes the RenderObject's
   /// semantics.
-  _SemanticsFragment _getSemanticsFragment() {
+  _SemanticsFragment _getSemanticsFragment({ bool mergeIntoParent: false }) {
     // early-exit if we're not dirty and have our own semantics
     if (!_needsSemanticsUpdate && isSemanticBoundary) {
       assert(_semantics != null);
-      return new _CleanSemanticsFragment(renderObjectOwner: this, dropSemanticsOfPreviousSiblings: isBlockingSemanticsOfPreviouslyPaintedNodes);
+      if (mergeIntoParent == _semantics.isMergedIntoParent) {
+        return new _CleanSemanticsFragment(renderObjectOwner: this, dropSemanticsOfPreviousSiblings: isBlockingSemanticsOfPreviouslyPaintedNodes);
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/object.dart]: probably wrap this too

File: packages/flutter/lib/src/rendering/semantics.dart [hunk 3]
```diff
@@ -341,17 +341,26 @@ class SemanticsNode extends AbstractNode with DiagnosticableTreeMixin {
...
+  bool _isMergedIntoParent = false;
+  set isMergedIntoParent(bool value) {
     assert(value != null);
-    if (_inheritedMergeAllDescendantsIntoThisNodeValue == value)
+    if (_isMergedIntoParent == value)
       return;
-    _inheritedMergeAllDescendantsIntoThisNodeValue = value;
+    _isMergedIntoParent = value;
     _markDirty();
   }
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/semantics.dart]: should these bools go in the flags?

File: packages/flutter/test/widgets/semantics_merge_test.dart [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+        children: <TestSemantics>[
+          new TestSemantics.rootChild(id: 1, label: 'test1'),
+          new TestSemantics.rootChild(id: 2, label: 'test2'),
+        ],
+      ),
+      ignoreRect: true,
+      ignoreTransform: true,
+    ));
+
+    //merged
```
Comment (Hixie) [inline packages/flutter/test/widgets/semantics_merge_test.dart]: most trivial of nits: missing space

File: packages/flutter/lib/src/rendering/object.dart [hunk 5]
```diff
@@ -2513,6 +2529,8 @@ abstract class RenderObject extends AbstractNode with DiagnosticableTreeMixin im
   /// determine if a node is previous to this one.
   bool get isBlockingSemanticsOfPreviouslyPaintedNodes => false;
 
+  bool get isMergingSemanticsOfDescendants => false;
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/object.dart]: you're trying to sneak an undocumented member into the codebase without me noticing! :-P

**Comment by Hixie** [comment]
I'm not a fan of isMergingSemanticsOfDescendants joining isBlockingSemanticsOfPreviouslyPaintedNodes in making the RenderObject API have all these Semantics-specific features, but I don't really have a better suggestion.

LGTM

File: packages/flutter/lib/src/rendering/object.dart (hunk 5 above)
Comment (goderbauer) [reply packages/flutter/lib/src/rendering/object.dart]: It was an honest mistake! I swear! 😆 

On a more serious note: It looks like we've documented all but one public member in the Flutter code base. Can we document the last remaining one and make the analyzer fail if new undocumented members are added? Or is something blocking that step?

File: packages/flutter/lib/src/rendering/semantics.dart (hunk 3 above)
Comment (goderbauer) [reply packages/flutter/lib/src/rendering/semantics.dart]: They are currently not transfered to the engine side and only processed on the framework side. That's why I've kept them separate from the flags for now.

**Comment by goderbauer** [comment]
> I'm not a fan of isMergingSemanticsOfDescendants joining isBlockingSemanticsOfPreviouslyPaintedNodes in making the RenderObject API have all these Semantics-specific features, but I don't really have a better suggestion.

I'd like to join that fan club.

**Pull request merged**